### PR TITLE
Basic Testing Tools

### PR DESCRIPTION
**Pull request by jayclassless**
I've added a few tools to make testing with `pytest` a little easier. They are slight rearrangements of what @andreypopp had put together for one of our other internal projects. They were helpful there, and are proving to be helpful in other projects. So, I'm pulling them upstream for easier reuse.

I initially thought to put them in a new packages (e.g., `rex.testing`), but instead of adding _yet another package_ to the platform, I decided to tuck them in with the packages they're most related to. Suggestions welcome.

### CODE REVIEW

File: src/rex.core/src/rex/core/testing.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        cls.rex = cls.create_app()
+        cls.rex.on()
+
+    @classmethod
+    def teardown_class(cls):
+        cls.rex.off()
+        cls.rex = None
+
+
+class any:
```
Comment (andreypopp) [inline src/rex.core/src/rex/core/testing.py]: Hm... I think I don't understand how this is going to be used. Isn't `assert x == any()` will be pass for any `x`, what's the assertion is doing then?

File: src/rex.core/src/rex/core/testing.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        my_variable = 'foo'
+        assert my_variable == any()  # True
+    """
+
+    def __eq__(self, other):
+        return True
+
+
+class not_none:
```
Comment (andreypopp) [inline src/rex.core/src/rex/core/testing.py]: II think `assert x is None` will work as well, no?

**Review by andreypopp — APPROVED** [review]
Thanks! Looks good to me.

I like that we are putting it inside existing packages and not creating new ones.

I don't understand the purpose of `any`/`none` though.

Also I personally prefer to write tests using "functional" API of pytest (https://github.com/prometheusresearch/baseline-codebase/blob/baseline/master/src/rex.graphql/test/test_graphql.py#L35), I guess it doesn't matter much at the end, just a note that some parts of the codebase use different style.

File: src/rex.core/src/rex/core/testing.py (hunk 1 above)
Comment (jayclassless) [reply src/rex.core/src/rex/core/testing.py]: The example is perhaps a bit simplistic, but this comes in handy when you're asserting an entire dictionary at once. E.g.:

```
assert something == {
  'foo': 1,
  'bar': any(),
  'baz': 'red',
}
```

This lets you assert the structure of the entire dictionary more clearly than X separate statements. You check that the dictionary contains exactly the keys you expect, but if there's one key whose value you don't care about what it specifically is, you can use `any()` as the placeholder.

File: src/rex.core/src/rex/core/testing.py (hunk 2 above)
Comment (jayclassless) [reply src/rex.core/src/rex/core/testing.py]: Yup, but see response above for `any()`.

File: src/rex.core/src/rex/core/testing.py (hunk 1 above)
Comment (andreypopp) [reply src/rex.core/src/rex/core/testing.py]: Oh, that's useful!

**Comment by jayclassless** [comment]
> Also I personally prefer to write tests using "functional" API of pytest (https://github.com/prometheusresearch/baseline-codebase/blob/baseline/master/src/rex.graphql/test/test_graphql.py#L35), I guess it doesn't matter much at the end, just a note that some parts of the codebase use different style.

Yea, this may be partially a matter of preference. We could make a fixture that would do these kinds of things for a functional approach, but I worry that the weight of instantiating a Rex app for every single case could make things mighty slow.

File: src/rex.core/src/rex/core/testing.py (hunk 2 above)
Comment (xitology) [reply src/rex.core/src/rex/core/testing.py]: I'm a bit concerned about masking the builtin function `any()`. How about `anything()`?

**Comment by andreypopp** [comment]
> We could make a fixture that would do these kinds of things for a functional approach, but I worry that the weight of instantiating a Rex app for every single case could make things mighty slow.

`fixture(scope='module')` makes it instantiate rex app only once per module

File: src/rex.core/src/rex/core/testing.py (hunk 2 above)
Comment (jayclassless) [reply src/rex.core/src/rex/core/testing.py]: That's fair, I'll rename.

**Comment by xitology** [comment]
Could you possibly reexport the new definitions from the package's `__init__.py`, and also add them to `doc/reference.rst`?

**Comment by jayclassless** [comment]
> Could you possibly reexport the new definitions from the package's `__init__.py`, and also add them to `doc/reference.rst`?

We could if we really want to, I didn't do it initially because then it'd put a hard dependency on `pytest`, which didn't seem appropriate. Would you be opposed to just making reference to these tools in the documentation, but not including them in `__init__.py`?

File: src/rex.core/src/rex/core/testing.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        cls.rex = cls.create_app()
+        cls.rex.on()
+
+    @classmethod
+    def teardown_class(cls):
+        cls.rex.off()
+        cls.rex = None
+
+
+class anything:
```
Comment (andreypopp) [inline src/rex.core/src/rex/core/testing.py]: Thinking more about this: probably `not_none` is useful in much more cases than `anything` cause `anything` is a really loose assertion (I don't remember I needed to check that key exists and don't do any assertion at least about its "nullability").

Still `anything` is useful when you don't really care about some part of the structure your asserting about.

Therefore I suggest:
* Rename `anything` => `ignored`
* Rename `not_none` => `anything_not_none`

Comment (jayclassless) [reply src/rex.core/src/rex/core/testing.py]: Renamed to `ignored` and `anything_except_none`.

**Comment by xitology** [comment]
> > Could you possibly reexport the new definitions from the package's `__init__.py`, and also add them to `doc/reference.rst`?
> 
> We could if we really want to, I didn't do it initially because then it'd put a hard dependency on `pytest`, which didn't seem appropriate. Would you be opposed to just making reference to these tools in the documentation, but not including them in `__init__.py`?

Even if we don't reexport them, pytest becomes a dependency of these packages and should be added to `setup.py`? Not that I like it. Although technically, `RexTestSuite` does not even depend on pytest, but `IsolatedCasesRexTestSuite` does.

By the way, what's the difference between `pytest.fixture(autouse=True)` and `setup_method()`/`teardown_method()`? Mixing `pytest.fixture()` with `setup_*`/`teardown_*` seems odd to me. Is it really necessary?

**Comment by jayclassless** [comment]
> By the way, what's the difference between `pytest.fixture(autouse=True)` and `setup_method()`/`teardown_method()`? Mixing `pytest.fixture()` with `setup_*`/`teardown_*` seems odd to me. Is it really necessary?

Good point. Not necessary at all. Solves the dependency issue.

I added these tools to the docs, and exported them from the packages (though, I kept them contained under the `testing` module, rather than littering the main packages with the individual objects).

**Review by xitology — APPROVED** [review]

**Pull request merged**